staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -893,3 +895,209 @@ func (m *mockTxnStore) Transaction(txns ...Transaction) *TransactionError {
 	return nil
 }
+
+func TestReplaceEvents(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	source := fcache.NewFakeControllerSource()
+	store := NewStore(DeletionHandlingMetaNamespaceKeyFunc)
+	t.Cleanup(func() {
+		source.Shutdown()
+	})
+
+	recorder := &eventRecorder{
+		updateCh: make(chan bool, 1),
+	}
+	fifo := NewRealFIFOWithOptions(RealFIFOOptions{
+		KeyFunction:  MetaNamespaceKeyFunc,
+		KnownObjects: store,
+	})
+
+	cfg := &Config{
+		Queue:            fifo,
+		ListerWatcher:    source,
+		ObjectType:       &v1.Pod{},
+		FullResyncPeriod: 0,
+
+		Process: func(obj interface{}, isInInitialList bool) error {
+			if deltas, ok := obj.(Deltas); ok {
+				return processDeltas(recorder, store, deltas, isInInitialList)
+			}
+			return errors.New("object given as Process argument is not Deltas")
+		},
+		ProcessBatch: func(deltaList []Delta, isInInitialList bool) error {
+			return processDeltasInBatch(recorder, store, deltaList, isInInitialList)
+		},
+	}
+
+	c := New(cfg)
+	go c.RunWithContext(ctx)
+	if !WaitForCacheSync(ctx.Done(), c.HasSynced) {
+		t.Fatal("Timed out waiting for cache sync")
+	}
+	testReplaceEvents(t, ctx, fifo, recorder, store)
+}
+
+func testReplaceEvents(t *testing.T, ctx context.Context, fifo Queue, m *eventRecorder, store Store) {
+	// Set up 3 pods initially
+	pKeep := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-keep", Namespace: "default"}}
+	pMod := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-mod", Namespace: "default", Labels: map[string]string{"ver": "1"}}}
+	pDel := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-del", Namespace: "default"}}
+
+	addFifo(t, fifo, pKeep)
+	addFifo(t, fifo, pMod)
+	addFifo(t, fifo, pDel)
+
+	// Wait for all 3 events to appear
+	err := m.waitForEventCount(ctx, 3, 5*time.Second)
+	if err != nil {
+		t.Fatalf("Controller failed to receive all initial setup events, got: %v", m.getHistory())
+	}
+
+	// Clear history so we only compare what happens during the RESYNC/REPLACE
+	m.clearHistory()
+
+	// Silent Drift (Events missed by controller)
+
+	// Modify: Update pod
+	pModUpdated := pMod.DeepCopy()
+	pModUpdated.Labels["ver"] = "2"
+
+	// Create pod-add
+	pAdd := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-add", Namespace: "default"}}
+
+	// Run replace
+	replaceFifo(t, fifo, []interface{}{pAdd, pKeep, pModUpdated})
+
+	err = m.waitForEventCount(ctx, 4, 5*time.Second)
+	if err != nil {
+		t.Fatalf("Controller failed to receive all events, got: %v", m.getHistory())
+	}
+
+	expectItems := []any{
+		pAdd,
+		pKeep,
+		pModUpdated,
+	}
+	items := sortItemsByKey(store.List(), DeletionHandlingMetaNamespaceKeyFunc)
+	if !apiequality.Semantic.DeepEqual(expectItems, items) {
+		t.Error(cmp.Diff(expectItems, items))
+	}
+	expectHistory := []eventRecord{
+		{Action: "add", Key: "default/pod-add"},
+		{Action: "delete", Key: "default/pod-del"},
+		{Action: "update", Key: "default/pod-keep"},
```

Comment:
Food for thought as a future idea.  If replace just just triggered based on watch resynchronization there is technically no reason to trigger update for it if the RV is matching. There is a small change that value was corrupted and we wanted to refresh it in controller, so we might need to design a separate mechanism to detect that, but it would be much cheaper for controllers.